### PR TITLE
TRI-156/Transaction Handling and Error Management for Project Integration

### PR DESCRIPTION
**Pull request by luccar31**
# Pull Request

## Description
Moved all transactional logic to the repository, refactor integration mechanism to support retrying against certain type of errors and email sending in cases where errors persists even after retrying 3 times.

## Ticket Link
[TRI-156](https://linear.app/tricker-sirius/issue/TRI-168/transaction-handling-and-error-management-for-project-integration)

### CODE REVIEW

File: src/domains/integration/service/integration.service.impl.ts [hunk 1]
```diff
@@ -65,120 +40,37 @@ export class IntegrationServiceImpl implements IntegrationService {
...
-
-        return newProject;
-      },
-      {
-        maxWait: 1000, // default: 2000
-        timeout: 20000, // default: 5000
-      }
-    );
+    Logger.time('Provider project data adaptation adaptation');
+    const projectData: ProjectDataDTO = await this.adapter.adaptProjectData({ providerProjectId: projectId, pmEmail: integratorEmail, token: pendingProject.token });
```
Comment (Juan-Bianchi) [inline src/domains/integration/service/integration.service.impl.ts]: Check logger message in line 51

File: src/domains/integration/service/integration.service.impl.ts [hunk 2]
```diff
@@ -65,120 +40,37 @@ export class IntegrationServiceImpl implements IntegrationService {
...
+    Logger.timeEnd(`Provider project data adaptation adaptation`);
+    await this.verifyPmExistence(projectData.members, integratorEmail);
```
Comment (Juan-Bianchi) [inline src/domains/integration/service/integration.service.impl.ts:85]: Please add return type explicitly, it is easier to follow the flow while checking code

File: src/domains/integration/repository/integration.repository.impl.ts [hunk 3]
```diff
@@ -46,4 +47,155 @@ export class IntegrationRepositoryImpl implements IntegrationRepository {
     });
     return new PendingProjectAuthorizationDTO(result);
   }
```
Comment (Juan-Bianchi) [inline src/domains/integration/repository/integration.repository.impl.ts:49]: Please add documentation

File: src/domains/integration/repository/integration.repository.impl.ts [hunk 4]
```diff
@@ -46,4 +47,155 @@ export class IntegrationRepositoryImpl implements IntegrationRepository {
...
+                userEmitter:
+                  event.userEmitterEmail != null
+                    ? {
+                        connect: {
+                          email: event.userEmitterEmail,
+                        },
+                      }
+                    : undefined,
+                eventRegisteredAt: event.userEmitterEmail,
+                from: event.from?.toString(),
```
Comment (Juan-Bianchi) [inline src/domains/integration/repository/integration.repository.impl.ts:193]: check line 186 please

File: src/domains/event/dto/index.ts [hunk 5]
```diff
@@ -1,8 +1,6 @@
 import { IsNotEmpty, IsString } from 'class-validator';
 
 export type BlockEventType = 'NO_STATUS' | 'BLOCKED_BY' | 'BLOCKING_TO';
-export type EventInputArray<T extends EventInput> = T[];
-
```
Comment (Juan-Bianchi) [inline src/domains/event/dto/index.ts:5]: Please document all classes modified

File: src/domains/adapter/dto/index.ts [hunk 6]
```diff
@@ -4,7 +4,6 @@ export interface AdaptProjectDataInputDTO {
   providerProjectId: string;
```
Comment (Juan-Bianchi) [inline src/domains/adapter/dto/index.ts:4]: Feel free to erase the 'DTO' part in the interface name if it is not being sent, recieved or returned by Database.

File: src/domains/integration/service/integration.service.impl.ts (hunk 2 above)
Comment (luccar31) [reply src/domains/integration/service/integration.service.impl.ts:85]: done

File: src/domains/integration/service/integration.service.impl.ts (hunk 1 above)
Comment (luccar31) [reply src/domains/integration/service/integration.service.impl.ts]: done

File: src/domains/integration/repository/integration.repository.impl.ts (hunk 4 above)
Comment (luccar31) [reply src/domains/integration/repository/integration.repository.impl.ts:193]: done

File: src/domains/event/dto/index.ts (hunk 5 above)
Comment (luccar31) [reply src/domains/event/dto/index.ts:5]: done

File: src/domains/adapter/dto/index.ts (hunk 6 above)
Comment (luccar31) [reply src/domains/adapter/dto/index.ts:4]: done

File: tests/domains/integration/service/integration.service.impl.test.ts [hunk 7]
```diff
@@ -115,8 +122,8 @@ describe('Integration service', () => {
       jest.spyOn(db, '$transaction').mockResolvedValue(project);
```
Comment (Juan-Bianchi) [inline tests/domains/integration/service/integration.service.impl.test.ts]: should erase mock in line 122

File: src/domains/email/service/mailgun.email.service.ts [hunk 8]
```diff
@@ -2,14 +2,22 @@ import { type IMailgunClient } from 'mailgun.js/Interfaces';
...
 import { type AuthorizationEmailVariables, type IntegrationConfirmationEmailVariables, type IntegrationRequestEmailVariables } from '@domains/email/dto';
-// import path from 'path';
 import { prepareHtmlTemplate } from '@utils/templating';
 import { HTML_TEMPLATES } from '@domains/email/service/templates';
 
-// TODO: fix file reads, delete hardcoded html strings
 export class MailgunEmailService implements EmailService {
   constructor(private readonly client: IMailgunClient) {}
 
+  async sendIntegrationErrorEmail(emailAddress: string): Promise<void> {
```
Comment (Juan-Bianchi) [inline src/domains/email/service/mailgun.email.service.ts:16]: please, document method

File: src/domains/event/dto/index.ts [hunk 9]
```diff
@@ -19,50 +17,64 @@ export enum LinearActionTypeConvention {
   REMOVE = 'r',
 }
 
-export class EventInput {
-  readonly providerEventId: string | null;
-  issueId: string;
-  readonly userEmitterEmail?: string;
-  readonly createdAt: Date;
-  userEmitterId?: string | null;
+export class EventData {
```
Comment (Juan-Bianchi) [inline src/domains/event/dto/index.ts:23]: Please remember to document classes

**Review by Juan-Bianchi — APPROVED** [review]

File: src/domains/email/service/mailgun.email.service.ts (hunk 8 above)
Comment (luccar31) [reply src/domains/email/service/mailgun.email.service.ts:16]: done

File: tests/domains/integration/service/integration.service.impl.test.ts (hunk 7 above)
Comment (luccar31) [reply tests/domains/integration/service/integration.service.impl.test.ts]: done

File: src/domains/event/dto/index.ts (hunk 9 above)
Comment (luccar31) [reply src/domains/event/dto/index.ts:23]: done

**Pull request merged**